File: backend/src/controllers/IncidentController.ts
import { Request, Response } from 'express';
import Incident from '../models/Incident.js';
import { Server as SocketServer } from 'socket.io';

export class IncidentController {
  private io: SocketServer;

  constructor(io: SocketServer) {
    this.io = io;
  }

  // Get all incidents
  getIncidents = async (req: Request, res: Response) => {
    try {
      const incidents = await Incident.find().sort({ createdAt: -1 });
      res.json(incidents);
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  };

  // Create an incident manually
  createIncident = async (req: Request, res: Response) => {
    try {
      const { orderId, driverId, type, severity, message } = req.body;

      if (!type || !message) {
        return res.status(400).json({ error: 'Type and message are required.' });
      }

      const incidentId = 'INC-' + Math.floor(100000 + Math.random() * 900000);

      const newIncident = new Incident({
        incidentId,
        orderId: orderId || null,
        driverId: driverId || null,
        type,
        severity: severity || 'medium',
        message,
        status: 'open'
      });

      await newIncident.save();
      this.io.emit('INCIDENT_CREATED', newIncident);

      res.status(201).json(newIncident);
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  };

  // Resolve an incident
  resolveIncident = async (req: Request, res: Response) => {
    try {
      const { incidentId } = req.params;

      const incident = await Incident.findOne({ incidentId });
      if (!incident) {
        return res.status(404).json({ error: 'Incident not found.' });
      }

      incident.status = 'resolved';
      await incident.save();

      this.io.emit('INCIDENT_RESOLVED', incident);
      res.json(incident);
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  };
}

File: backend-java/src/main/java/com/fleet/dispatch/controller/IncidentController.java
package com.fleet.dispatch.controller;

import com.fleet.dispatch.model.Incident;
import com.fleet.dispatch.repository.IncidentRepository;
import com.fleet.dispatch.websocket.SocketIOService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/incidents")
@CrossOrigin(origins = "*")
public class IncidentController {

    @Autowired
    private IncidentRepository incidentRepository;

    @Autowired
    private SocketIOService socketIOService;

    @GetMapping
    public ResponseEntity<List<Incident>> getIncidents() {
        return ResponseEntity.ok(incidentRepository.findAllByOrderByTimestampDesc());
    }

    @PostMapping
    public ResponseEntity<?> createIncident(@RequestBody Map<String, String> body) {
        String type = body.get("type");
        String message = body.get("message");

        if (type == null || message == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Type and message are required."));
        }

        String incidentId = "INC-" + (int)(100000 + Math.random() * 900000);
        Incident incident = new Incident(
                incidentId,
                body.get("orderId"),
                body.get("driverId"),
                type,
                body.getOrDefault("severity", "medium"),
                message
        );

        Incident saved = incidentRepository.save(incident);
        socketIOService.emit("INCIDENT_CREATED", saved);

        return ResponseEntity.status(HttpStatus.CREATED).body(saved);
    }

    @PutMapping("/{incidentId}/resolve")
    public ResponseEntity<?> resolveIncident(@PathVariable String incidentId) {
        Optional<Incident> incOpt = incidentRepository.findByIncidentId(incidentId);
        if (incOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Incident not found."));
        }

        Incident inc = incOpt.get();
        inc.setStatus("resolved");
        Incident saved = incidentRepository.save(inc);

        socketIOService.emit("INCIDENT_RESOLVED", saved);
        return ResponseEntity.ok(saved);
    }
}
